### PR TITLE
Update assemblyscript. Cleanup as code. Add build script

### PR DESCRIPTION
**Pull request by MaxGraey**

### CODE REVIEW

**Comment by magcius** [comment]
What's the goal of this PR?

**Comment by magcius** [comment]
You also did not run update_wat_modules.js so this has no effect.

File: src/assembly/gx_texture_as.ts [hunk 1]
```diff
@@ -86,16 +77,13 @@ export function decode_I8(pScratch: u32, pDst: u32, pSrc: u32, w: u32, h: u32):
...
-    const bh: u8 = 4;
-
     let srcOffs: u32 = 0;
-    for (let yy: u32 = 0; yy < h; yy += bh) {
-        for (let xx: u32 = 0; xx < w; xx += bw) {
-            for (let y: u8 = 0; y < bh; y++) {
-                for (let x: u8 = 0; x < bw; x++) {
-                    let dstPixel: u32 = (w * (yy + y)) + xx + x;
-                    let dstOffs: u32 = pDst + dstPixel * 4;
+    for (let yy: u32 = 0; yy < h; yy += 4) {
```
Comment (magcius) [inline src/assembly/gx_texture_as.ts]: Can the AssemblyScript compiler not do this constant folding? I don't like this manual folding.

**Comment by MaxGraey** [comment]
It's just proposal. You can accept or decline. But I tried speedup your routines a little

File: src/assembly/gx_texture_as.ts [hunk 2]
```diff
@@ -222,21 +198,21 @@ export function decode_RGBA8(pScratch: u32, pDst: u32, pSrc: u32, w: u32, h: u32
 @inline
 function s3tcblend(a: u32, b: u32): u8 {
     // return (a*3 + b*5) / 8;
-    let tmp: u32 = (((a << 1) + a) + ((b << 2) + b)) >>> 3;
-    return <u8> tmp;
+    let tmp = (((a << 1) + a) + ((b << 2) + b)) >>> 3;
+    return <u8>tmp;
```
Comment (magcius) [inline src/assembly/gx_texture_as.ts]: Unless https://github.com/AssemblyScript/assemblyscript/issues/104 was solved, this still needs the explicit u32 cast AFAIUI.

File: src/assembly/gx_texture_as.ts (hunk 1 above)
Comment (MaxGraey) [reply src/assembly/gx_texture_as.ts]: Currently LICM pass disabled by binaryen so short answer - no

File: src/assembly/gx_texture_as.ts (hunk 2 above)
Comment (MaxGraey) [reply src/assembly/gx_texture_as.ts]: Why? `a` and `b` has same types

Comment (magcius) [reply src/assembly/gx_texture_as.ts]: I remember having a lot of difficulty with this. But if you've tested it and it works, then fine.

**Comment by magcius** [comment]
Please run update_wat_modules.js and test this. If you need help getting set up with the game data, please DM me on Twitter (@JasperRLZ)

File: src/assembly/gx_texture_as.ts (hunk 2 above)
Comment (MaxGraey) [reply src/assembly/gx_texture_as.ts]: if you don't trust me. See this fiddle. It's pretty the same)
https://webassembly.studio/?f=bty2wv7g5jr

**Comment by MaxGraey** [comment]
Ok, I will test this tomorrow. Currently mark this as WIP

File: src/assembly/gx_texture_as.ts (hunk 1 above)
Comment (MaxGraey) [reply src/assembly/gx_texture_as.ts]: btw LICM pass should be done by browsers VMs but currently even simple loop rotation is highly experimental on some browsers like Chrome. So better optimize loops stuffs manually

**Comment by MaxGraey** [comment]
@magcius Where is `update_wat_modules.js` located?

**Comment by magcius** [comment]
Sorry, it's called `built_wat.js`: https://github.com/magcius/noclip.website/blob/master/src/build_wat.js

**Comment by MaxGraey** [comment]
```
/Volumes/Archive/Projects/Github/noclip.website/src/build_wat.js:62
    const wabtModule = wabt.parseWat(watFilename, wat);
                            ^

TypeError: wabt.parseWat is not a function
```

**Comment by magcius** [comment]
Hence my old yarn.lock files. Sorry, I don't have time to chase after all the API breaks in JavaScript land. I really don't want to sound dismissive, and I genuinely appreciate the help, but you're going to have to debug this one on your own :)

**Comment by MaxGraey** [comment]
ok, no problem

**Comment by MaxGraey** [comment]
Hmm, it seems script always fail in this part:
```ts
export function decompress(srcBuffer: ArrayBufferSlice): Promise<ArrayBufferSlice> {
    return _wasmInstance.then((wasmInstance) => {
        const srcView = srcBuffer.createDataView();
        assert(readString(srcBuffer, 0x00, 0x04) === 'Yaz0'); // <---- fail
        ...
```
What could it be?

**Comment by MaxGraey** [comment]
I tried upstream version and this exactly the same issue. On latest Chrome and Firefox.

**Comment by magcius** [comment]
Do you have the ROM data in the correct location? It tries to load off the local disk by default.

**Comment by MaxGraey** [comment]
I just run `start` script and that's all

**Comment by magcius** [comment]
Modify this to use the hosted storage: https://github.com/magcius/noclip.website/blob/master/src/fetch.ts#L11

**Comment by MaxGraey** [comment]
Hmm, I commented 3 lines and forcing load from server but got 403 status =(

**Comment by magcius** [comment]
Sorry about that -- we're currently in the process of migrating the storage bucket. There's an older version of the data up at https://z.noclip.website/ which might work for some games but not others. Super Mario Sunshine, for instance, should work.

**Comment by MaxGraey** [comment]
Unfortunately ` https://z.noclip.website/` also not works due CORS. Or, will wait until you finish migration process. Let me know when this happened

**Comment by magcius** [comment]
The migration might take a while -- ETA is still unknown. I just added CORS support to z.noclip.website. Sorry about this.

**Comment by MaxGraey** [comment]
Thanks. It seems all works as expected. So I remove **WIP** status. I could make more improvements to AS code if you want

File: src/assembly/Yaz0_as.ts [hunk 3]
```diff
@@ -13,32 +15,33 @@ function set(offs: u32, b: u8): void {
 
 @inline
 function get16be(offs: u32): u16 {
-    return (load<u8>(offs) << 8) | load<u8>(offs + 1);
+    // return (load<u8>(offs) << 8) | load<u8>(offs + 1);
+    return bswap(load<u16>(offs));
```
Comment (magcius) [inline src/assembly/Yaz0_as.ts]: What's the reason for this change?

File: src/build_wat.js [hunk 4]
```diff
@@ -40,31 +39,15 @@ function buildModuleExportsInterface(exportName, wasmModule) {
     return lines.join('\n');
 }
 
-function compileAssemblyScript(filename) {
-    const outFilename = filename.replace('.ts', '.wat');
-    // TODO(jstpierre): noRuntime flag is gone, noLib seems to cause an assert fail in the compiler somewhere... ugh.
-    const res = spawnSync('node', [path.join(__dirname, '../node_modules/assemblyscript/bin/asc'), '--baseDir', __dirname, filename, '--noLib', '-O', '-c', '-o', outFilename]);
```
Comment (magcius) [inline src/build_wat.js]: Any reason for removing this in favor of the npm script?

File: src/wat_modules.ts [hunk 5]
```diff
@@ -3,8 +3,9 @@
 // assembly/Yaz0_as.ts
 export interface yaz0_asExports {
     memory: WebAssembly.Memory;
+    table: WebAssembly.Table;
```
Comment (magcius) [inline src/wat_modules.ts]: The addition of the table makes me scared. I rely on a very specific memory layout. What is this for?

Comment (MaxGraey) [reply src/wat_modules.ts]: It just workaround because you code throw exception when found unknown export symbol. Anyway this not used so you shouldn't worry about it

File: src/build_wat.js (hunk 4 above)
Comment (MaxGraey) [reply src/build_wat.js]: because npm script more flexible and clear. Also you din't need wabt because AS generate wat and wabt simultaneously so don't need extra transform step `wat -> wasm`

File: src/assembly/Yaz0_as.ts (hunk 3 above)
Comment (MaxGraey) [reply src/assembly/Yaz0_as.ts]: One read memory operation instead 2. That's always better and faster

Comment (magcius) [reply src/assembly/Yaz0_as.ts]: OK. Please remove the commented out line then.

Comment (MaxGraey) [reply src/assembly/Yaz0_as.ts]: sure!

**Comment by magcius** [comment]
I'll be looking at this in an hour or so. Please let me know when you feel it's ready for merge.

**Comment by MaxGraey** [comment]
It basically ready

File: src/assembly/gx_texture_as.ts [hunk 6]
```diff
@@ -101,10 +113,11 @@ export function decode_IA4(pScratch: u32, pDst: u32, pSrc: u32, w: u32, h: u32):
                     let ia: u8 = get(pSrc + srcOffs);
                     let a: u8 = expand4to8(ia >>> 4);
                     let i: u8 = expand4to8(ia & 0x0F);
-                    set(dstOffs + 0, i);
+                    /*set(dstOffs + 0, i);
                     set(dstOffs + 1, i);
                     set(dstOffs + 2, i);
-                    set(dstOffs + 3, a);
+                    set(dstOffs + 3, a);*/
+                    set32(dstOffs, (<u32>i * 0x00010101) | (<u32>a << 24));
```
Comment (magcius) [inline src/assembly/gx_texture_as.ts]: Is this guaranteed to be little-endian?

File: src/assembly/gx_texture_as.ts [hunk 7]
```diff
@@ -76,10 +87,11 @@ export function decode_I8(pScratch: u32, pDst: u32, pSrc: u32, w: u32, h: u32):
                     let dstPixel = stride2 + x;
                     let dstOffs = pDst + dstPixel * 4;
                     let i = get(pSrc + srcOffs);
-                    set(dstOffs + 0, i);
+                    /*set(dstOffs + 0, i);
                     set(dstOffs + 1, i);
                     set(dstOffs + 2, i);
-                    set(dstOffs + 3, i);
+                    set(dstOffs + 3, i);*/
+                    set32(dstOffs, <u32>i * 0x01010101);
```
Comment (magcius) [inline src/assembly/gx_texture_as.ts]: Heh, clever trick. However, please remove commented-out code.

File: src/assembly/gx_texture_as.ts (hunk 6 above)
Comment (MaxGraey) [reply src/assembly/gx_texture_as.ts]: It should store pixel data in `AARRGGBB` / `AABBGGRR` or `AALLLLLL` in current case. Hard to tell. Also will be great have some tests for that. I tested visually in dfferent ROMs and it seems fine, but I'm not sure is this rom used that texture format or not)

Comment (magcius) [reply src/assembly/gx_texture_as.ts]: The current output of the script is R, G, B, A, as four separate bytes in increasing offsets. I don't know what set32 does -- does it assume the memory order is little-endian? I'm assuming this comes down to WebAssembly's memory set operation?

Comment (MaxGraey) [reply src/assembly/gx_texture_as.ts]: Yes it little-endian always

Comment (MaxGraey) [reply src/assembly/gx_texture_as.ts]: https://github.com/WebAssembly/design/blob/master/Portability.md#assumptions-for-efficient-execution

**Comment by magcius** [comment]
Thank you for your contribution! Sorry about the harsh review.

**Comment by magcius** [comment]
Hi, ever since this landed, I've started to see "WebAssembly: out of memory" errors after refreshing the page a lot. Would you have any idea as to why this now happens?

**Comment by MaxGraey** [comment]
Hmm, I don't see this issue on Firefox 66b12 and Chrome 73. And that's strangle because you allocate and control memory by host. That's not "out of bounds" error it's exactly "out of memory" which could mean you have memory leaks somewhere on host side. Is it dev (watch) mode or production?

**Comment by MaxGraey** [comment]
Also I saw this warning on Firefox:
```
Error: WebGL warning: drawArrays: Drawing without vertex attrib 0 array enabled forces 
the browser to do expensive emulation work when running on desktop OpenGL platforms, 
for example on Mac. It is preferable to always draw with vertex attrib 0 array enabled, 
by using bindAttribLocation to bind some always-used attribute to location 0.
```

**Comment by magcius** [comment]
Sorry. I should have mentioned that I saw that error on Chrome, and only after a *large* amount of refreshing and development. I saw it when working on OoT3D object spawns, and hot reload activates a lot, so it might have been something like 50 times.

Mind telling me which game you see that WebGL performance warning on?

**Comment by MaxGraey** [comment]
> Mind telling me which game you see that WebGL performance warning on?
On any platform or game. But this print only once at web app startup or after refresh

**Comment by magcius** [comment]
I'm definitely not seeing it here:
image

**Comment by MaxGraey** [comment]
Mat be this only on Mac or only on Firefox 66 beta? Hmm

noclip

**Comment by magcius** [comment]
Possibly fixed by 06cbb9a8ba7d5aa5d8b11e0e89f6ca238b53e4a6 -- try again now?

**Comment by MaxGraey** [comment]
I tried noclip's site. Cleared cache but still see this warning. Or I need wait little bit more?

**Comment by magcius** [comment]
No, it should be live. Bizarre that it didn't fix it. A vertex attribute of 0 should always be bound now. I unfortunately don't have any Mac devices to test on, so I can't reproduce.

**Comment by MaxGraey** [comment]
@magcius It seems OOM which you mentioned is general problem for latest Chrome:
https://github.com/brave/brave-browser/issues/3366

**Comment by magcius** [comment]
Hm. According to those comments it happens because SharedArrayBuffer isn't collected, but I'm not using SAB in my codebase. Unless WebAssembly.Memory is a SAB?

**Comment by MaxGraey** [comment]
Also: https://bugs.chromium.org/p/chromium/issues/detail?id=935169
